pandas/core/sorting.py

```diff
@@ -52,7 +52,21 @@ def _int64_cut_off(shape):
                 return i
         return len(shape)
 
-    def loop(labels, shape):
+    def maybe_lift(lab, size):
+        # promote nan values (assigned to -1 here)
+        # so that all values are non-negative
+        return (lab + 1, size + 1) if (lab == -1).any() else (lab, size)
```

Comment:
Does this not obfuscate NA and 0 values?

Reply:
this works with labels, not the original values. labels got here from 
df.duplicate -> core.algorithms.factorize -> _factorize_array call and that method replace NA with -1 (`na_sentinel : int, default -1`)
Also, I did not change this in any way, it's an existing code that already in master :)